README.txt
-----------------------------------------------------------------------------------
Project: OpenGTS - Open GPS Tracking System
URL    : http://www.opengts.org
File   : README.txt
-----------------------------------------------------------------------------------

Notes:
 - Please refer to the included "OpenGTS_Config.pdf" file for all installation and 
   configuration information for the OpenGTS system.
 - When updating to a newer version of OpenGTS, it is highly recommended that the 
   following command be run to update the database with the latest table/column 
   changes:
     $  bin/dbAdmin.pl -tables=ca
   Or, run the following if alternate key fields also need to be updated:
     $  bin/dbAdmin.pl -tables=cak
 - On Windows 7, Internet Explorer 8, if you receive the error 
      Error [jsMapInit:[object error]
   then you may need to enable "Active Scripting" in the security tab.
 - A listing of the database tables and fields can be found in the file "SCHEMA.txt"
   in the OpenGTS installation diretory.
 - Documentation is available online at the following link:
     http://www.opengts.org/documentation.html
 - This open-source version of OpenGTS includes support for the following GPS 
   tracking devices:
     - Aspicore GSM Tracker (Nokia, Samsung, Sony Ericsson phones).
       [http://www.aspicore.com]
     - Astra Telematics AT240, AT110, AT210.
       [http://www.gps-telematics.co.uk]
     - Sanav GC-101, MT-101, and CT-24 Personal Tracker (HTTP-based protocol).
       Sanav GX-101 Vehicle Tracker (HTTP-based protocol).
       [http://www.sanav.com]
     - V-Sun 3338 Personal Tracker.
     - Android App "GPS2OpenGTS"
       [http://www.androidzoom.com/android_applications/communication/gps2opengts_nfsk_download.html]
     - iCare G3300 Personal Tracker.
     - Most TK102/TK103 devices.
     - Xexun TK102.
     - ZhongShan SIPGEAR Technology Co, Ltd.
     - TAIP (Trimble ASCII Interface Protocol).
     - TrackStick GPS data logger.
     - "GPSMapper" capable phones.
     - "NetGPS" capable devices.
     - HP hw6965 Windows/CE phone (OpenDMTP compliant)
     - Mologogo capable phones
     - Various Boost Mobile phones (OpenDMTP compliant)

Please let us know if you find any issues with this release.

-----------------------------------------------------------------------------------

Visit the following link for additional downloadable documentation and add-ons for 
OpenGTS:
  http://www.opengts.org/documentation.html
  http://www.geotelematic.com/documentation.html

-----------------------------------------------------------------------------------

This package may contain one or more of the following source modules which contain 
copyrights from their respective authors.  Please review these source modules for 
their copyright and license information:
   war/track/js/mapstraction/mapstraction.js
   war/track/js/mapstraction/mapstraction-geocode.js
   war/track/js/mapstraction/labeledmarker.js
   war/track/js/sorttable/sorttable.js

-----------------------------------------------------------------------------------

Contributors:
  See the following link for a list of contributors:
    http://www.opengts.org/info.html

-----------------------------------------------------------------------------------

Sample data:
Additional documentation for installing sample 'demo' data into the database can be
found in the "README.txt" file in the "sampleData" directory at
    sampleData/README.txt

-----------------------------------------------------------------------------------

Boost Mobile Motorola phones:
For information regarding support for various Boost Mobile Motorola phones, please
see the document at 
    MotoDMTP/MotoDMTP.txt

-----------------------------------------------------------------------------------

Sanav GC-101 GPS Tracking Device Support:
Additional documentation for installing and configuring the GC-101 server within
OpenGTS can be found in the "README.txt" file in the "gc101" source directory at
   src/org/opengts/war/gc101/README.txt

See the following link for manufacturer's product information:
   http://www.sanav.com

-----------------------------------------------------------------------------------

TAIP (Trimble ASCII Interface Protocol) Support:
Support for TAIP (Trimble ASCII Interface Protocol) is included in this release.  
This server uses the raw-socket mode device communication server based on the example
'template' server.  The "taip.jar" server jar file can be built with the command:
   > ant taip
And can be started in the same manner that other servers are started using the
"runserver.pl" command as follows:
   > $GTS_HOME/bin/runserver.pl -s taip

-----------------------------------------------------------------------------------

ICare G3300 Personal GPS Tracking Device Support:
Support for the ICare G3300 GPS tracking device is included in this release.  This
server uses the raw-socket mode device communication server based on the example
'template' server.  The "icare.jar" server jar file can be built with the command:
   > ant icare
And can be started in the same manner that other servers are started using the
"runserver.pl" command as follows:
   > $GTS_HOME/bin/runserver.pl -s icare

-----------------------------------------------------------------------------------

Updating existing Geozone table entries 
(only required when upgrading from versions previous to v1.9.3):

If you have existing Geozone table entries, they will need to be updated to include
Bounding-Box information used to optimize Geozone lookups.  This update process is
done automatically with the following steps:

1) Update the new Geozone bounding-box table fields:
     > bin/dbAdmin.pl -tables=ca

2) For each account having Geozone entries, run the following command (replace "<accountID>" 
   with the appropriate account-IDs for each account which has Geozone entries):
     > bin/admin.pl Geozone -account=<accountID> -list -update

The update process should now be complete.

-----------------------------------------------------------------------------------

Runtime config file property key references
(for *.conf files, such as 'default.conf', 'common.conf', 'webapp.conf, ...):
- The prefixing "#" is a comment character, indicating that the following text is for
  informational purposes only, and should not be parsed.  Comments must be placed on
  a separate line, and must not be specified at the end of a property specification.
- Previously defined property keys may be referenced by placing them in ${...} brackets,
  such as ${log.name}.  Environment variables may also be referenced in this manner, as in 
  ${GTS_HOME}, or ${CATALINA_HOME}.
- A default value may be specified for referenced property keys which have not been defined.   
  For instance ${OUTPUT_DIR=/tmp} would resolve to "/tmp", if the property OUTPUT_DIR is not 
  defined (as an environment variable, or otherwise), or will resolve to the value specified 
  on the OUTPUT_DIR property (or environment variable) if it is defined.
- Recursive propery key references may also be specified.  For instance, the specification
  ${THIS_DIR=${THAT_DIR=/tmp}} will first attempt to resolve the value for "THIS_DIR", if not
  found, then value of THAT_DIR will be returned.  If THAT_DIR is not defined, then finally
  the specification will resolve to "/tmp".
- Property reference specifications for which the property is not defined, and there is no
  specified default value will resolve to the literal "${var}" string.  That is, if "var"
  is not defined, then ${var} will resolve to the literal string "${var}".
- Property reference specifications are only resolved at the time a property value is 
  requested, not at the time the property key=value line is parsed (ie. lazy resolution).  
  For instance, in the property specification below, "var_a" will resolve to "test", even 
  though "var_b" is not defined until after "var_a":
    var_a=${var_b}
    var_b=test
  This means that if "var_b" should ever change, the resolved value of "var_a" will also
  change accordingly.  [ie. "var_a" will only be resolved at the time a call to an RTConfig
  property value retrieval method is called - such as RTConfig.getString("var_a")].
  Write-only (assignable) property keys (ie. %log, %include) are the only exception to this 
  rule.  Since their assigned behavior is executed at the time they are parsed ('%log' prints
  a log message, and '%include' includes a file), write-only property keys must be assigned
  values that can be fully resolved at the time their values are assigned.
- Property keys may be re-assigned.  Property keys take on their last assigned value.  For
  instance, the following will display "Hello", then display "World":
    test=Hello
    %log=${test}
    test=World
    %log=${test}
    
The following are reserved read-only constant property keys:
   %version        returns the current version, as in "2.3.4"
   %contextName    returns the context 'name' (name of the servlet context, or main class name)
   %contextPath    returns the context 'path' (path to the servlet context, or $GTS_HOME)
   %configURL      returns the URL for the loaded config file
   %hostName       returns the current host-name
   %hostIP         returns the current mail host IP address

The following are reserved write-only (assignable) constant property keys:
   %log            displays the specified value to the log output
   %include        includes the config at the value URL (URL must exist, error otherwise)
   %include?       includes the config at the value URL (URL may exist, otherwise ignored)
   (The included URL protocol must be one of "file" or "http".)
   
The following are available within a PrivateLabel session context:
   session.name    The PrivateLabel Domain name
   session.locale  The PrivateLabel Domain locale

-----------------------------------------------------------------------------------

All trade names listed above are trademarks of their respective companies.

CHANGELOG.txt
-----------------------------------------------------------------------------------
GTS Changelog
Copyright 2007-2012 GeoTelematic Solutions, Inc.
-----------------------------------------------------------------------------------

Version 2.4.7 [2012/12/27]
- NEW: Added additional status codes.
- NEW: Added "priority" column to EventData table ("CustomFieldInfo" optional fields)
- FIX: Fix Reverse-Geocoding for cell-tower locations when RG 'alwaysFast' is true.
- FIX: Added proper sort ordering for temperature fields in reports.
- FIX: Removed "ORDER BY" fields when retrieving only a record "COUNT(*)".
- DCS: gprmc] If 'mobileID' and 'accountID' are blank, use 'deviceID' as 'mobileID'.
- DCS: astra] New Astra DCS provided by Astra Telematics.

Version 2.4.6 [2012/11/02]
- CHG: Changed EntityManager TRAILER type to '1'.
- NEW: Added "entityType" field to EventData table ("CustomFieldInfo" option fields)
- NEW: Added custom outbound SMS gateway error checking.
- NEW: Added additional status codes.
- NEW: Added hardcoded support to "ReportSpreadsheet.java" for numeric columns.
- NEW: Added new "-rtpidfile=" command line option to write the current PID file.
- DCS: tk10x] Added "LOAD" and "ON" acknowledgements, where required.
- FIX: ReportBody methods "getRecordCount" and "isPartial" are now thread-safe.
- FIX: Changed SQL Server data types for UINT8/INT8, UINT16/INT16.

Version 2.4.5 [2012/09/02] 
- CHG: Increased number of geozone points to 10.
- NEW: Added additional columns to EventData table.
- NEW: Added a "maxMessageLength" property to "SMSOutboundGateway" support.
- NEW: Added "createAge" column option to report layout "EventDataLayout".

Version 2.4.4 [2012/08/01] 
- NEW: Additional pushpins, special thanks to Jason Herring.
- NEW: Added option to Account table for finding referenced email addresses.
- NEW: Added additional fields to Device table.
- NEW: Added ${vehicleVolts} message replacement variable to EventData.
- NEW: Added "reportDistance" column option to report layout "EventDataLayout".
- NEW: Added "Assigned User ID" option to Device/Vehicle Admin page.
- NEW: Added "lastUpdateUser" column support to various tables.
- DCS: gprmc] Fixed NPE when no date/time is specified.

Version 2.4.3 [2012/06/29]
- NEW: Additional status codes.
- CHG: Added event 'createAge' to map pushpin events sent to the client browser.
- DCS: New Lantrix T1800/T2000 DCS.  Provided by Carlos Ramos.
- FIX: Check for divide-by-zero in Account when EconomyUnits is set to "L/100km".

Version 2.4.2 [2012/05/27]
- NEW: Added command-line "-delete" option to "Geozone.java".
- NEW: Added check for validating local host name to "CheckInstall".
- FIX: Fixed NPE issue in EventData/Device/Account modules when adding optional
       fields to the map info-balloon with "OptionalEventFields.DeviceMap=".
- FIX: Fixed connection pool issue that could cause a memory leak when using
       DB connection pooling thanks to Brijraj Vaghani).
- DCS: aspicore - added simulated geozones and estimated odometer calculations.
       GPS location is invalid.
       display if enabled.

Version 2.4.1 [2012/04/20]
- NEW: Added additional defined Status Codes.
- NEW: Ability to delete old events by DeviceGroup (as well as by Device). Also
       checks the very last event when deleting old events.  If the last event is
       prior to the deleted time range, the last event will be retained.
- NEW: Added additional columns types to EventDataLayout.java and FieldLayout.java.
- NEW: Added ability to hide "Unique ID", "SIM#", on Device Admin page via ACLs.
- NEW: Added "userInfo.showAccessControlList" property to hide ACL list on User Admin.
- NEW: Added field "simID" to Device record.
- NEW: Added command-line option to delete old EventData records by device group.
- NEW: Added command-line support for displaying "events-per-second" in EventData.
- NEW: Made GISGraphy street address lookup optional.
- NEW: Added Mapquest OSM/Arial lay support to OpenLayers MapProvider support. (See
       "openLayers" MapProvider definition for details.) Make sure to comply with 
       their terms of service.
- NEW: Added "Event Count by Day" device detail report.
- DCS: gprmc - update to be able to handle several protocol formats simultaneously.
- DCS: gc101 - fixed NPE issue when invalid $GPRMC record is parsed.
- DCS: template - "parseInsertRecord_ASCII_02" now allows using AccountID as MobileID.
- DCS: tk10x - added digital input change simulated events.
- FIX: Fixed typo - changed property key "groupInfo.deviceLisSortBy" name to 
       "groupInfo.deviceListSortBy".

Version 2.4.0 [2012/02/05]
- NEW: Added additional defined Status Codes.
- NEW: Added initial support for sending SMS messages from the OpenGTS
       "RuleFactoryExample.java" module (SMS outbound gateway config required).
- NEW: Added Economy-Units "L/100km" (Litres per 100 kilometers).
- NEW: Added "lastIgnitionOffTime" column to Device table.
- NEW: CheckInstall now shows report limits, SMS gateway configuration.

war/track/js/TrackMap.js
// ----------------------------------------------------------------------------
// Copyright 2007-2012, GeoTelematic Solutions, Inc.
// All rights reserved
// ----------------------------------------------------------------------------
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ----------------------------------------------------------------------------
// Change History:
//  2008/08/17  Martin D. Flynn
//     -Moved from TrackMap.java
//  2008/09/19  Martin D. Flynn
//     -Added support for updating map with "Last" point.
//     -Added support for starting "AutoUpdate" on map load.
//  2009/04/02  Martin D. Flynn
//     -Added 'trackMapEventLimit' to allow event 'limit' overrides.
//  2009/11/01  Martin D. Flynn
//     -If "AutoUpdateOnLoad" is true, clicking "Update" now just resets/restarts
//      the auto-update timer.
//  2009/11/10  Martin D. Flynn
//     -Added condition to only restart auto-update if "AutoUpdateOnLoad" is true
//      AND auto-update is currently active/in-process.
// ----------------------------------------------------------------------------

var ID_DEVICE_ID        = "deviceSelector";
var ID_DEVICE_DESCR     = "deviceDescription";
var ID_GOTO_ADDRESS     = "gotoAddressFld";

var PAGE_ZONEGEOCODE    = "ZONEGEOCODE";
var TAG_geocode         = "geocode";
var TAG_geonames        = "geonames";
var TAG_code            = "code";
var TAG_geoname         = "geoname";
var TAG_lat             = "lat";
var TAG_lng             = "lng";
var TAG_lon             = "lon";

// ----------------------------------------------------------------------------

var trackMapLimitOverride = -1;

/* return the event 'limit' for non-fleet maps */
function trackMapEventLimit()
{
    return IS_FLEET? -1 : trackMapLimitOverride;
};

// ----------------------------------------------------------------------------

/* this is executed when the page is loaded */
function trackMapOnLoad()
{

    /* display calendars */
    if (mapCal_fr) {
        mapCal_fr.setCollapsible(CalendarCollapsible, CalendarFade, CalendarDivBox);
        mapCal_to.setCollapsible(CalendarCollapsible, CalendarFade, CalendarDivBox);
        calWriteCalendars(mapCal_fr, mapCal_to);
        if ((CalendarDateOnLoad == "last") && jsvLastEventYMD && (jsvLastEventEpoch > 0)) {
            mapCal_to.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, jsvLastEventYMD.DD);
            mapCal_fr.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, jsvLastEventYMD.DD);
        }
    } else {
        mapCal_to.writeCalendar();
        if ((CalendarDateOnLoad == "last") && jsvTodayYMD) {
            mapCal_to.setDate(jsvTodayYMD.YYYY, jsvTodayYMD.MM, jsvTodayYMD.DD);
        }
    }

    /* init AutoUpdate button text */
    var btn = document.getElementById(ID_MAP_AUTOUPDATE_BTN);
    if (btn != null) { btn.value = AutoUpdateEnable? TEXT_autoUpdateStart : "??"; }

    /* collapse map controls */
    if (ControlCollapseOnLoad) {
        jsmControlDisplay(false);
    }

    /* update map */
    if (AutoUpdateOnLoad) {
        // start AutoUpdate (first RecenterMode should be RECENTER_ZOOM)
        var savedRecenterMode = AutoUpdateRecenterMode;
        AutoUpdateRecenterMode = jsmRecenterZoomMode(RECENTER_ZOOM);
        //trackMapClickedAutoUpdate();
        startAutoUpdateMapTimer();
        AutoUpdateRecenterMode = savedRecenterMode;
    } else {
        // update map points
        var limit = ((MapUpdateOnLoad == "last") && !IS_FLEET)? 1 : trackMapEventLimit();
        trackMapUpdateMap(limit, trackMapLimitType(), jsmRecenterZoomMode(RECENTER_ZOOM), 0);
    }

};

/* this is executed when the page is unloaded */
function trackMapOnUnload()
{
    mapProviderUnload();
};

// ----------------------------------------------------------------------------

/* this is executed when a new region is selected */
function trackMapSelectRegion()
{

    /* remove current shapes */
    jsmDrawShape("!");

    /* find zoomRegion */
    var zoomRegion = null;
    if (trackZoomRegionShapes) {
        var zrName = document.SelectDeviceForm.regionSelector.value;
        if ((zrName != "") && (zrName != "normal")) {
            for (var i = 0; i < trackZoomRegionShapes.length; i++) {
                if (zrName == trackZoomRegionShapes[i].name) {
                    zoomRegion = trackZoomRegionShapes[i];
                    break;
                }
            }
        }
    }
    if (zoomRegion == null) {
        // selected zoom region not found
        jsmSetFixedZoom(false);
        trackMapClickedUpdateAll();
        return;
    }
    
    /* attributes */
    var color = zoomRegion.color;
    var type  = zoomRegion.type;   // numParseFloat(fld[0], 0);
    var radM  = zoomRegion.radius; // numParseFloat(fld[1], 0);

    // ZoomRegion value:
    // <type>,<radius>,<lat>/<lon>,<lat>/<lon>
    var fld = zoomRegion.points.split(',');
    //if (fld.length < 3) { return; }
    //var type  = numParseFloat(fld[0], 0);
    //var radM  = numParseFloat(fld[1], 0);
    var jspt  = [];
    for (var i = 0; i < fld.length; i++) {
        var LL = fld[i].split('/');
        if (LL.length < 2) { continue; }
        var lat = numParseFloat(LL[0], 0);
        var lon = numParseFloat(LL[1], 0);
        if (((lat != 0) || (lon != 0))) {
            jspt.push(new JSMapPoint(lat,lon));
        }
    }
    if (jspt.length == 0) {
        jsmSetFixedZoom(false);
        trackMapClickedUpdateAll();
        return;
    }
    
    /* legacy zoom mode */
    if ((type == "zoom") || (type == "!zoom")) {
        jsmSetCenter(jspt[0].lat, jspt[0].lon, radM);
        jsmSetFixedZoom(true);
        return;
    }

    /* draw shape */
    if (jsmDrawShape(type, radM, jspt, color, true)) {
        jsmSetFixedZoom(true);
        return;
    }
    
    /* not supported */
    jsmSetFixedZoom(false);

};

// ----------------------------------------------------------------------------

/* reset the calendar dates to the last known device date */
function _resetCalandarDates()
{
    if (!IS_FLEET && jsvLastEventYMD && (jsvLastEventEpoch > 0)) {
        if (mapCal_to) {
            var day = jsvLastEventYMD.DD + 1;  // next day
            mapCal_to.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, day);
        }
        if (mapCal_fr) { 
            var day = jsvLastEventYMD.DD;  // this day
            mapCal_fr.setDate(jsvLastEventYMD.YYYY, jsvLastEventYMD.MM, day); 
        }
    } else
    if (jsvTodayYMD) {
        if (mapCal_to) {
            var day = jsvTodayYMD.DD + 1;  // next day
            //alert("_resetCalandarDates: YYYY="+jsvTodayYMD.YYYY+ " MM="+jsvTodayYMD.MM+ " DD="+day);
            mapCal_to.setDate(jsvTodayYMD.YYYY, jsvTodayYMD.MM, day);
        }
        if (mapCal_fr) { 
            var day = jsvTodayYMD.DD;  // this day
            mapCal_fr.setDate(jsvTodayYMD.YYYY, jsvTodayYMD.MM, day); 
        }
    } else {
        // as a last resort, we could rely on the 'Date' returned by the browser
        // (which may be inaccurate), but for now just leave the calendars as-is.
    }
};

/* this is executed when the "Latest Event" is selected */
function trackMapGotoLastEventDate()
{
    _resetCalandarDates();
    trackMapClickedUpdateAll();
};

// ----------------------------------------------------------------------------

/* this is executed when the "Update Map"/"Auto Update" is pressed */
function trackMapPingDevice(devcmd) 
{
    if (!IS_FLEET && mapDevicePing && DEVICE_PING_URL && (DEVICE_PING_URL != "")) { 
        var dev = document.SelectDeviceForm.device.value; // PARM_DEVICE
        var rfr = mapCal_fr.getArgDateTime();
        var rto = mapCal_to.getArgDateTime();
        var tmz = calGetTimeZone();
        var url = DEVICE_PING_URL + 
            "&_uniq=" + Math.random() +  // necessary to make the URL unique
            "&" + PARM_RANGE_FR + "=" + strEncode(rfr) +        // necessary?
            "&" + PARM_RANGE_TO + "=" + strEncode(rto) +        // necessary?
            "&" + PARM_TIMEZONE + "=" + strEncode(tmz) +        // necessary?
            "&" + PARM_DEVICE_GROUP + "=" + strEncode(dev) +    
            "&" + PARM_DEVICE_COMMAND + "=" + strEncode(devcmd?devcmd:"");
        mapDevicePing(url); // valid iff 'jsmap.js' is in use!
    }
};

// ----------------------------------------------------------------------------

/* get limit type first/last */
function trackMapLimitType()
{
    return LimitType;
};

/* this is executed when the "Update All" is pressed */
function trackMapClickedUpdateAll() // all
{
    if (AutoUpdateOnLoad && (AutoUpdateMapTimer != null)) {
        startAutoUpdateMapTimer();
    } else {
        stopAutoUpdateMapTimer();
        trackMapUpdateMap(trackMapEventLimit(), trackMapLimitType(), jsmRecenterZoomMode(RECENTER_ZOOM), 0);
    }
};

/* this is executed when the "Update Last" is pressed */
function trackMapClickedUpdateLast() // last
{
    stopAutoUpdateMapTimer();
    trackMapUpdateMap(1, "last", jsmRecenterZoomMode(RECENTER_ZOOM), 0);
};

/* this is executed when the "Replay Map" is pressed */
function trackMapClickedReplay(showInfoBox) 
{
    //try { document.getElementById(ID_CENTER_LAST_POINT_FORM).centerLastPoint.checked = false; } catch (e) {}
    stopAutoUpdateMapTimer();
    var replay = showInfoBox? 2 : 1; // 1=don't show info box, 2=show info box (if enabled)
    var replayState = mapProviderPauseReplay(replay);
    if (replayState > 0) {
        // replay paused/resumed
    } else {
        // (re)starting replay
        trackMapUpdateMap(trackMapEventLimit(), trackMapLimitType(), jsmRecenterZoomMode(RECENTER_ZOOM), replay);
    }
};

/* this is executed when "Update Map" is clicked */
function trackMapUpdateMap(limit, limitType, recenterMode, replay) 
{
    var limitFirst = false;
    //try { document.getElementById(ID_CENTER_LAST_POINT_FORM).centerLastPoint.checked = false; } catch (e) {}
    var dev = IS_FLEET? document.SelectDeviceForm.group.value : document.SelectDeviceForm.device.value; // PARM_GEOUP/PARM_DEVICE
    var rfr = mapCal_fr? mapCal_fr.getArgDateTime() : "";
    var rto = mapCal_to.getArgDateTime();
    var tmz = calGetTimeZone();
    var url = MAP_UPDATE_URL + 
        "&_uniq=" + Math.random() +  // necessary to make the URL unique
        "&" + PARM_RANGE_FR + "=" + strEncode(rfr) +
        "&" + PARM_RANGE_TO + "=" + strEncode(rto) +
        "&" + PARM_TIMEZONE + "=" + strEncode(tmz) +
        "&" + PARM_DEVICE_GROUP + "=" + strEncode(dev);
    if (limit > 0) {
        url += "&" + PARM_LIMIT + "=" + limit;
    }
    if (limitType != "") {
        url += "&" + PARM_LIMIT_TYPE + "=" + limitType;
    }
    mapProviderUpdateMap(url, recenterMode, replay);
    // "Replay" button ID is ID_MAP_REPLAY_BTN
};

/* this is executed when "Update KML" is clicked */
function trackMapUpdateKML() 
{
    if (KML_UPDATE_URL && (KML_UPDATE_URL != "")) {
        var dev = IS_FLEET? document.SelectDeviceForm.group.value : document.SelectDeviceForm.device.value; // PARM_GEOUP/PARM_DEVICE
        var rfr = mapCal_fr? mapCal_fr.getArgDateTime() : "";
        var rto = mapCal_to.getArgDateTime();
        var tmz = calGetTimeZone();
        var url = KML_UPDATE_URL +  // Google KML
            "&_uniq=" + Math.random() +  // necessary to make the URL unique
            "&" + PARM_RANGE_FR + "=" + strEncode(rfr) +
            "&" + PARM_RANGE_TO + "=" + strEncode(rto) +
            "&" + PARM_TIMEZONE + "=" + strEncode(tmz) +
            "&" + PARM_DEVICE_GROUP + "=" + strEncode(dev);
        openURL(url, "_blank");
    }
};

/* this is executed when the "Last Location" is pressed */
//function trackMapClickedCenterOnLast()
//{
//    stopAutoUpdateMapTimer();
//    try { jsmCenterOnLastPushpin(document.getElementById(ID_CENTER_LAST_POINT_FORM).centerLastPoint.checked); } catch (e) {/*ignore*/}
//};

// ----------------------------------------------------------------------------

/* auto-update map (toggle) */
function trackMapClickedAutoUpdate()
{
    if (AutoUpdateMapTimer != null) {
        stopAutoUpdateMapTimer();
    } else {
        startAutoUpdateMapTimer();
    }
};

/* sets the AutoUpdateMap button text */
var _autoUpdateButtonElem = null;
function _setAutoUpdateButtonText(text)
{
    if (_autoUpdateButtonElem == null) {
        _autoUpdateButtonElem = document.getElementById(ID_MAP_AUTOUPDATE_BTN);
    }
    if (_autoUpdateButtonElem != null) { 
        _autoUpdateButtonElem.value = text; 
    }
};

/* periodic map update timer target */
function _timerAutoUpdateMap() 
{
    if (--AutoIntervalCount <= 0) {
        _resetCalandarDates();
        var limit = ((MapUpdateOnLoad == "last") && !IS_FLEET)? 1 : trackMapEventLimit();
        trackMapUpdateMap(limit, "last", jsmRecenterZoomMode(AutoUpdateRecenterMode), 0);
        if ((AutoMaxCount > 0) && (++AutoUpdateMapCount >= AutoMaxCount)) {
            // we've reached the maximum number of allowed updates.
            stopAutoUpdateMapTimer();
            return;
        }
        AutoIntervalCount = AutoInterval; // start over
    }
    _setAutoUpdateButtonText(TEXT_autoUpdateStop + ' : ' + AutoIntervalCount);
};

/* start a map auto-update timer */
function startAutoUpdateMapTimer() 
{
    stopAutoUpdateMapTimer();
    _resetCalandarDates();
    var limit = ((MapUpdateOnLoad == "last") && !IS_FLEET)? 1 : trackMapEventLimit(); // single last point, or all points
    trackMapUpdateMap(limit, "last", jsmRecenterZoomMode(AutoUpdateRecenterMode), 0); // update map now
    AutoIntervalCount  = AutoInterval;
    AutoUpdateMapCount = 0;
    _setAutoUpdateButtonText(TEXT_autoUpdateStop);
    AutoUpdateMapTimer = setInterval('_timerAutoUpdateMap()',1000); // setTimeout
};

/* stop any running map auto-update timer */
function stopAutoUpdateMapTimer()
{
    if (AutoUpdateMapTimer != null) {
        clearInterval(AutoUpdateMapTimer); // clearTimeout
        AutoUpdateMapTimer = null;
    }
    AutoIntervalCount  = 0;
    AutoUpdateMapCount = 0;
    _setAutoUpdateButtonText(TEXT_autoUpdateStart);
};

// ----------------------------------------------------------------------------

/* show device selector */
// only valid if DeviceChooser has been included
function trackMapShowSelector()
{
    if (deviceShowChooserList) {
        var list = (typeof TrackSelectorList != 'undefined')? TrackSelectorList : null;
        deviceShowChooserList(ID_DEVICE_ID, ID_DEVICE_DESCR, list); 
    }
};

function deviceDeviceChanged()
{
    trackMapSelectDevice();
};

/* this is executed when a new device is selected */
function trackMapSelectDevice()
{ // 'PARM_PAGE=map' && PARM_DEVICE are defined in FORM_NEW_DEVICE
    document.SelectDeviceForm.date_fr.value = mapCal_fr? mapCal_fr.getArgDateTime() : ""; // PARM_RANGE_FR
    document.SelectDeviceForm.date_to.value = mapCal_to.getArgDateTime(); // PARM_RANGE_TO
    document.SelectDeviceForm.date_tz.value = calGetTimeZone(); // PARM_TIMEZONE
    document.SelectDeviceForm.submit();
};

// ----------------------------------------------------------------------------

function trackMapGotoAddress()
{
    var addrFld = document.getElementById(ID_GOTO_ADDRESS);
    if (addrFld == null) { return; }
    var addr = addrFld.value;
    var ct   = "US";

    /* get the latitude/longitude for the zip */
    var url = "./Track?page=" + PAGE_ZONEGEOCODE + "&addr=" + addr + "&country=" + ct + "&_uniq=" + Math.random();
    //alert("URL " + url);
    try {
        var req = jsmGetXMLHttpRequest();
        if (req) {
            req.open("GET", url, true);
            //req.setRequestHeader("CACHE-CONTROL", "NO-CACHE");
            //req.setRequestHeader("PRAGMA", "NO-CACHE");
            //req.setRequestHeader("If-Modified-Since", "Sat, 1 Jan 2000 00:00:00 GMT");
            req.onreadystatechange = function() {
                if (req.readyState == 4) {
                    var lat = 0.0;
                    var lon = 0.0;
                    for (;;) {
                        
                        /* get xml */
                        var xmlStr = req.responseText;
                        if (!xmlStr || (xmlStr == "")) {
                            break;
                        }
                        
                        /* get XML doc */
                        var xmlDoc = createXMLDocument(xmlStr);
                        if (xmlDoc == null) {
                            break;
                        }

                        /* try parsing as "geocode" encasulated XML */
                        var geocode = xmlDoc.getElementsByTagName(TAG_geocode);
                        if ((geocode != null) && (geocode.length > 0)) {
                            //alert("geocode: " + xmlStr);
                            var geocodeElem = geocode[0];
                            if (geocodeElem != null) {
                                var latn = geocodeElem.getElementsByTagName(TAG_lat);
                                var lonn = geocodeElem.getElementsByTagName(TAG_lng);
                                if (!lonn || (lonn.length == 0)) { lonn = geocodeElem.getElementsByTagName(TAG_lon); }
                                if ((latn.length > 0) && (lonn.length > 0)) {
                                    lat = numParseFloat(latn[0].childNodes[0].nodeValue,0.0);
                                    lon = numParseFloat(lonn[0].childNodes[0].nodeValue,0.0);
                                    break;
                                }
                            }
                            break;
                        }

                        /* try parsing as forwarded XML from Geonames */
                        var geonames = xmlDoc.getElementsByTagName(TAG_geonames);
                        if ((geonames != null) && (geonames.length > 0)) {
                            //alert("geonames: " + xmlStr);
                            // returned XML was forwarded as-is from Geonames
                            var geonamesElem = geonames[0];
                            var codeList = null;
                            if (geonamesElem != null) {
                                codeList = geonamesElem.getElementsByTagName(TAG_code);
                                if (!codeList || (codeList.length == 0)) {
                                    codeList = geonamesElem.getElementsByTagName(TAG_geoname);
                                }
                            }
                            if (codeList != null) {
                                for (var i = 0; i < codeList.length; i++) {
                                    var code = codeList[i];
                                    var latn = code.getElementsByTagName(TAG_lat);
                                    var lonn = code.getElementsByTagName(TAG_lng);
                                    if ((latn.length > 0) && (lonn.length > 0)) {
                                        lat = numParseFloat(latn[0].childNodes[0].nodeValue,0.0);
                                        lon = numParseFloat(lonn[0].childNodes[0].nodeValue,0.0);
                                        break;
                                    }
                                }
                            }
                            break;
                        }
                        
                        /* break */
                        //alert("unknown: " + xmlStr);
                        break;
                        
                    }

                    /* set lat/lon */
                    if ((lat != 0.0) || (lon != 0.0)) {
                        // got lat/lon
                        //alert("Lat="+lat +", Lon="+lon);
                        //jsmSetCenter(lat,lon,PUSHPIN_ZOOM);
                        jsmDrawShape("!");
                        jsmDrawShape("circle", 200, [ new JSMapPoint(lat,lon) ], "#888855", true);
                    }

                } else
                if (req.readyState == 1) {
                    // alert('Loading GeoNames from URL: [' + req.readyState + ']\n' + url);
                } else {
                    // alert('Problem loading URL? [' + req.readyState + ']\n' + url);
                }
            }
            req.send(null);
        } else {
            alert("Error [trackMapGotoAddress]:\n" + url);
        }
    } catch (e) {
        alert("Error [trackMapGotoAddress]:\n" + e);
    }

};

// ----------------------------------------------------------------------------
